### DISCUSSION TITLE
Optimize build/test execution time

### DISCUSSION THREAD

**Discussion by stephan-strate** [quarkiverse/quarkus-quinoa#154] [Ideas]
First of all: Thanks for your work. This extension is amazing :)

I would like to start a discussion about optimizing the execution time for testing and building the frontend application.
Problem: Having the backend and frontend application separately, each needing build time (including tests) of eg. 3 minutes. When using quinoa this build time increases to 6 minutes, because these processes are executed sequentially.

First idea: Wouldn't it be possible to execute backend and frontend tests in parallel?

What do you think? I am happy to help and contribute.

**Comment by ia3andy** [comment]
Hello @stephan-strate  :)

Thanks for the great feedback. It makes me really happy to hear those!

> First idea: Wouldn't it be possible to execute backend and frontend tests in parallel?

Something we could do: make it easy to call the webui tests from a specific Junit test, then, it's up to anyone to choose how they should be ran through JUnit (sequential or //).. It would also give a finer control over running those tests, we could even imagine allow to call different frontend test group from different junit methods (and have an environment/options to update the snapshots if used). Also, I think it makes sense to have those test closer to the Playwright E2E tests (if there are some).

> Optimizing the build time

For this, I am not sure we can do anything as it is tied to the Quarkus build system. I am all hear for propositions thought..


Cheers,

**Reply by stephan-strate** [reply]
> Something we could do: make it easy to call the webui tests from a specific Junit test, then, it's up to anyone to choose how they should be ran through JUnit (sequential or //).. It would also give a finer control over running those tests, we could even imagine allow to call different frontend test group from different junit methods (and have an environment/options to update the snapshots if used). Also, I think it makes sense to have those test closer to the Playwright E2E tests (if there are some).

That sounds really good. Gives a lot of freedom and control over the frontend things.
Maybe a dumb question: Do parallel tests work with Quarkus? Does it work with standard junit configuration? I could imagine that it does not work too good to execute such tests in parallel.

> For this, I am not sure we can do anything as it is tied to the Quarkus build system. I am all hear for propositions thought..

Yeah. I am not sure either. It would be great, but I have no idea how to tackle it.

**Reply by ia3andy** [reply]
@geoand might have some thoughts?

**Reply by geoand** [reply]
Parallel junit test execution is not advised in Quarkus, there are too many things that can go wrong

**Reply by ia3andy** [reply]
@stephan-strate So you will only be able to run the junit test which run the webui test in // to the other (need to check there is no port conflict), you'll still win a bit of time I guess. Do you think it's worth the effort?

**Reply by stephan-strate** [reply]
So eg. having 3 min backend tests + 3 min frontend tests, running them in parallel would be totally worth it in my opinion. I am not quite sure if it is worth it to execute them as junit tests though. Right now they are executed in the build process, right? Is it even possible to execute something async in there?

**Reply by ia3andy** [reply]
> So eg. having 3 min backend tests + 3 min frontend tests, running them in parallel would be totally worth it in my opinion. I am not quite sure if it is worth it to execute them as junit tests though. Right now they are executed in the build process, right? Is it even possible to execute something async in there?

We could start a separate thread for the webui tests, and re-join when the latest (build or test) is done.

**Reply by stephan-strate** [reply]
Do you have any advice on where to start? Would love to contribute to this.

**Reply by ia3andy** [reply]
Test run here:
https://github.com/quarkiverse/quarkus-quinoa/blob/main/deployment/src/main/java/io/quarkiverse/quinoa/deployment/QuinoaProcessor.java#L128

and build is the next one..

**Reply by stephan-strate** [reply]
I am trying to get my head around this.

Maybe we could get some good results very easy using multiple maven modules. Having the webui including tests in a separate maven module with no dependencies and running the maven command with `-T 2` should execute tests and builds of the modules with no dependencies in parallel.

I will test this theory and bring back the results. If it works, we can just put in the documentation.

**Reply by ia3andy** [reply]
@stephan-strate TBH I think it's worth the effort to try having two threads (build and test) in QuinoaProcessor, it would make things easier to setup for devs.

**Reply by stephan-strate** [reply]
@ia3andy Got it. But this would only speed up the frontend test/build. Backend and frontend would still run sequentially.

**Reply by ia3andy** [reply]
Ok, I get your point, so yeah you could give a try to the maven multi-modules to get an idea of how significant changing this would be. I am quite skeptical on the gain, I am eager to have your results :)

**Reply by stephan-strate** [reply]
@ia3andy I ran into some problems executing my idea. As the whole execution is tied to the `quarkus-maven-plugin`, I can not put the frontend into a separat maven module. As far as I know maven multi-module project always have one module which bundles all the others and has the `quarkus-maven-plugin` installed. All others do just use jandex to index the CDI beans.

**Reply by ia3andy** [reply]
@aloubyansky might be able to help you..

**Reply by aloubyansky** [reply]
If the goal is only to run tests, you don't need to configure the `quarkus-maven-plugin` in a module unless there are source code generating extensions in the project. What you need is `quarkus-junit5` in `test` scope.

**Reply by stephan-strate** [reply]
@aloubyansky `quarkus-quinoa` runs the frontend build and tests using a BuildStep. So as far as I understand, `quarkus-junit5` won't do anything here.
